## Higest Cost Order - Amazon/Shopify - Hard Category

In [1]:
import pandas as pd
import sqlite3
import sql
sql.style = 'plain'
%load_ext sql
connection=sqlite3.connect("question_bank.db")
print("Successfully connected to SQL database")
# Connect to SQLite
%sql sqlite:///question_bank.db
print("success")

Successfully connected to SQL database
success


In [10]:
%%sql
DROP TABLE IF EXISTS orders;

CREATE TABLE orders (
    id INT PRIMARY KEY,
    cust_id INT,
    order_date DATE,
    order_details VARCHAR(50),
    total_order_cost INT
);

-- 2. Insert data
INSERT INTO orders (id, cust_id, order_date, order_details, total_order_cost) VALUES
(1, 3, '2019-03-04', 'Coat', 100),
(2, 3, '2019-03-01', 'Shoes', 80),
(3, 3, '2019-03-07', 'Skirt', 30),
(4, 7, '2019-02-01', 'Coat', 25),
(5, 7, '2019-03-10', 'Shoes', 80),
(6, 15, '2019-02-01', 'Boats', 100),
(7, 15, '2019-01-11', 'Shirts', 60),
(8, 15, '2019-03-11', 'Slipper', 20),
(9, 15, '2019-03-01', 'Jeans', 80),
(10, 15, '2019-03-09', 'Shirts', 50),
(11, 5, '2019-02-01', 'Shoes', 80),
(12, 12, '2019-01-11', 'Shirts', 60),
(13, 12, '2019-03-11', 'Slipper', 20),
(14, 4, '2019-02-01', 'Shoes', 80),
(15, 4, '2019-01-11', 'Shirts', 60),
(16, 3, '2019-04-19', 'Shirts', 50),
(17, 7, '2019-04-19', 'Suit', 150),
(18, 15, '2019-04-19', 'Skirt', 30),
(19, 15, '2019-04-20', 'Dresses', 200),
(20, 12, '2019-01-11', 'Coat', 125),
(21, 7, '2019-04-01', 'Suit', 50),
(22, 7, '2019-04-02', 'Skirt', 30),
(23, 7, '2019-04-03', 'Dresses', 50),
(24, 7, '2019-04-04', 'Coat', 25),
(25, 7, '2019-04-19', 'Coat', 125),
(26, 3, '2019-04-20', 'Gloves', 20),
(27, 3, '2019-04-21', 'Tie', 25),
(28, 3, '2019-04-22', 'Cap', 15),
(29, 3, '2019-04-23', 'Jacket', 120),
(30, 1, '2019-04-19', 'Jacket', 150),
(31, 1, '2019-04-19', 'Shoes', 125);

 * sqlite:///question_bank.db
Done.
Done.
31 rows affected.


[]

In [11]:
%%sql
DROP TABLE IF EXISTS customers;

CREATE TABLE customers (
    id INT PRIMARY KEY,
    first_name VARCHAR(50),
    last_name VARCHAR(50),
    city VARCHAR(50),
    address VARCHAR(100),
    phone_number VARCHAR(20)
);

INSERT INTO customers (id, first_name, last_name, city, address, phone_number) VALUES
(8, 'John', 'Joseph', 'San Francisco', NULL, '928-386-8164'),
(7, 'Jill', 'Michael', 'Austin', NULL, '813-297-0692'),
(4, 'William', 'Daniel', 'Denver', NULL, '813-368-1200'),
(5, 'Henry', 'Jackson', 'Miami', NULL, '808-601-7513'),
(13, 'Emma', 'Isaac', 'Miami', NULL, '808-690-5201'),
(14, 'Liam', 'Samuel', 'Miami', NULL, '808-555-5201'),
(15, 'Mia', 'Owen', 'Miami', NULL, '808-640-5201'),
(1, 'Mark', 'Thomas', 'Arizona', '4476 Parkway Drive', '602-993-5916'),
(12, 'Eva', 'Lucas', 'Arizona', '4379 Skips Lane', '301-509-8805'),
(6, 'Jack', 'Aiden', 'Arizona', '4833 Coplin Avenue', '480-303-1527'),
(2, 'Mona', 'Adrian', 'Los Angeles', '1958 Peck Court', '714-409-9432'),
(10, 'Lili', 'Oliver', 'Los Angeles', '3832 Euclid Avenue', '530-695-1180'),
(3, 'Farida', 'Joseph', 'San Francisco', '3153 Rhapsody Street', '813-368-1200'),
(9, 'Justin', 'Alexander', 'Denver', '4470 McKinley Avenue', '970-433-7589'),
(11, 'Frank', 'Jacob', 'Miami', '1299 Randall Drive', '808-590-5201');

 * sqlite:///question_bank.db
Done.
Done.
15 rows affected.


[]

In [21]:
%%sql
select * from orders
order by cust_id
limit 5;

 * sqlite:///question_bank.db
Done.


[(30, 1, '2019-04-19', 'Jacket', 150),
 (31, 1, '2019-04-19', 'Shoes', 125),
 (1, 3, '2019-03-04', 'Coat', 100),
 (2, 3, '2019-03-01', 'Shoes', 80),
 (3, 3, '2019-03-07', 'Skirt', 30)]

In [22]:
%%sql
select * from customers
order by id
limit 2;

 * sqlite:///question_bank.db
Done.


[(1, 'Mark', 'Thomas', 'Arizona', '4476 Parkway Drive', '602-993-5916'),
 (2, 'Mona', 'Adrian', 'Los Angeles', '1958 Peck Court', '714-409-9432')]

In [34]:
%%sql
select count(distinct first_name)
from customers

 * sqlite:///question_bank.db
Done.


[(15,)]

* Find the customers with the highest daily total order cost between 2019-02-01 and 2019-05-01. 
* If a customer had more than one order on a certain day, sum the order costs on a daily basis. 
* Output each customer's first name, total cost of their items, and the date.
* For simplicity, you can assume that every first name in the dataset is unique.

In [44]:
%%sql

-- Lets join via cust_id and id

select
    c.first_name,
    o.order_date,
    o.order_details,
    o.total_order_cost
from customers c
join orders o
on c.id=o.cust_id
where order_date between '2019-02-01' and '2019-05-01'
order by o.order_date

 * sqlite:///question_bank.db
Done.


[('Jill', '2019-02-01', 'Coat', 25),
 ('Mia', '2019-02-01', 'Boats', 100),
 ('Henry', '2019-02-01', 'Shoes', 80),
 ('William', '2019-02-01', 'Shoes', 80),
 ('Farida', '2019-03-01', 'Shoes', 80),
 ('Mia', '2019-03-01', 'Jeans', 80),
 ('Farida', '2019-03-04', 'Coat', 100),
 ('Farida', '2019-03-07', 'Skirt', 30),
 ('Mia', '2019-03-09', 'Shirts', 50),
 ('Jill', '2019-03-10', 'Shoes', 80),
 ('Mia', '2019-03-11', 'Slipper', 20),
 ('Eva', '2019-03-11', 'Slipper', 20),
 ('Jill', '2019-04-01', 'Suit', 50),
 ('Jill', '2019-04-02', 'Skirt', 30),
 ('Jill', '2019-04-03', 'Dresses', 50),
 ('Jill', '2019-04-04', 'Coat', 25),
 ('Farida', '2019-04-19', 'Shirts', 50),
 ('Jill', '2019-04-19', 'Suit', 150),
 ('Mia', '2019-04-19', 'Skirt', 30),
 ('Jill', '2019-04-19', 'Coat', 125),
 ('Mark', '2019-04-19', 'Jacket', 150),
 ('Mark', '2019-04-19', 'Shoes', 125),
 ('Mia', '2019-04-20', 'Dresses', 200),
 ('Farida', '2019-04-20', 'Gloves', 20),
 ('Farida', '2019-04-21', 'Tie', 25),
 ('Farida', '2019-04-22', 'Cap', 15),
 ('Farida', '2019-04-23', 'Jacket', 120)]

* Tried to extract day and months - but it creates a lot of confusion.
* It's better to sort dates normally and sql will understand, it has date datatype
* no need to explain or extract, so this will not work also, but just tried

In [128]:
%%sql
select
    c.first_name,
    strftime('%m', order_date) AS month,
    strftime('%d', order_date) AS day,
    sum(o.total_order_cost)
from customers c
join orders o
on c.id=o.cust_id
where order_date between '2019-02-01' and '2019-05-01'
group by day
order by month,day
limit 5;

 * sqlite:///question_bank.db
Done.


[('Farida', '03', '01', 495),
 ('Farida', '03', '04', 125),
 ('Farida', '03', '07', 30),
 ('Mia', '03', '09', 50),
 ('Jill', '03', '10', 80)]

## Solution

In [132]:
%%sql
WITH daily_costs as (
    select
        c.first_name,
        o.order_date,
        sum(o.total_order_cost) as daily_total
    from customers c
    join orders o
    on c.id=o.cust_id
    where order_date between '2019-02-01' and '2019-05-01'
    group by c.first_name,o.order_date
),
ranked as (
    select *,
        row_number() over (partition by order_date order by daily_total desc) as rk
    from daily_costs
)
select * from ranked
where rk=1

 * sqlite:///question_bank.db
Done.


[('Mia', '2019-02-01', 100, 1),
 ('Farida', '2019-03-01', 80, 1),
 ('Farida', '2019-03-04', 100, 1),
 ('Farida', '2019-03-07', 30, 1),
 ('Mia', '2019-03-09', 50, 1),
 ('Jill', '2019-03-10', 80, 1),
 ('Eva', '2019-03-11', 20, 1),
 ('Jill', '2019-04-01', 50, 1),
 ('Jill', '2019-04-02', 30, 1),
 ('Jill', '2019-04-03', 50, 1),
 ('Jill', '2019-04-04', 25, 1),
 ('Jill', '2019-04-19', 275, 1),
 ('Mia', '2019-04-20', 200, 1),
 ('Farida', '2019-04-21', 25, 1),
 ('Farida', '2019-04-22', 15, 1),
 ('Farida', '2019-04-23', 120, 1)]

* ROW_NUMBER is dropping rows when there are ties (i.e., multiple customers with the same maximum daily spend).
* For example:
    * On 2019-03-01, both Mia (80) and Farida (80) should appear.
    * On 2019-03-11, both Mia (20) and Eva (20) should appear.
    * On 2019-04-19, both Jill (275) and Mark (275) should appear.
------------
* But your query using ROW_NUMBER() only keeps one row per day.

* To fix this, you should use RANK() (or DENSE_RANK()) instead of ROW_NUMBER() so ties are preserved.

In [133]:
%%sql
WITH daily_costs as (
    select
        c.first_name,
        o.order_date,
        sum(o.total_order_cost) as daily_total
    from customers c
    join orders o
    on c.id=o.cust_id
    where order_date between '2019-02-01' and '2019-05-01'
    group by c.first_name,o.order_date
),
ranked as (
    select *,
        rank() over (partition by order_date order by daily_total desc) as rk
    from daily_costs
)
select * from ranked
where rk=1

 * sqlite:///question_bank.db
Done.


[('Mia', '2019-02-01', 100, 1),
 ('Farida', '2019-03-01', 80, 1),
 ('Mia', '2019-03-01', 80, 1),
 ('Farida', '2019-03-04', 100, 1),
 ('Farida', '2019-03-07', 30, 1),
 ('Mia', '2019-03-09', 50, 1),
 ('Jill', '2019-03-10', 80, 1),
 ('Eva', '2019-03-11', 20, 1),
 ('Mia', '2019-03-11', 20, 1),
 ('Jill', '2019-04-01', 50, 1),
 ('Jill', '2019-04-02', 30, 1),
 ('Jill', '2019-04-03', 50, 1),
 ('Jill', '2019-04-04', 25, 1),
 ('Jill', '2019-04-19', 275, 1),
 ('Mark', '2019-04-19', 275, 1),
 ('Mia', '2019-04-20', 200, 1),
 ('Farida', '2019-04-21', 25, 1),
 ('Farida', '2019-04-22', 15, 1),
 ('Farida', '2019-04-23', 120, 1)]

* See, here - On 2019-03-01, both Mia (80) and Farida (80) appears